### PR TITLE
[MM-44954] Regenerate default avatar

### PR DESCRIPTION
**Pull request by thinkGeist**
#### Summary
This PR adds the ability to regenerate the default profile picture for a user when their username is changed. 

**NOTE: This PR has been recreated from #20544 to resolve branch conflicts with the mono-repo change, there are some pre-existing discussions there**

<details>
<summary>See a video preview of the altered behaviour here</summary>

Current state of the change in this PR:

https://user-images.githubusercontent.com/9254315/225044448-04bc848f-e274-4d30-bafc-c64a42505b99.mp4

</details>

In order to support this operation an adjustment needed to be made to how we determine a user has a default profile picture. Before this change this was done by checking `lastPictureUpdate == 0`, this change alters the logic to instead be `lastPictureUpdate <= 0` to support updating client applications with the new profile picture automatically when the username is changed. 

To support automagically updating the profile picture for a user on the client application it is necessary that that `lastPictureUpdate` field has changed from the client's previously known value when a `user_updated` websocket event is encountered. 
In order to support this instead of setting `lastPictureUpdate` to `0` when `ResetLastPictureUpdate` is called we instead set it to the current time (in milliseconds since epoch) cast to a negative number. This way we can still easily differentiate between a default profile image and one set by a user, but retain the ability to detect "updates" to this field when regenerating the default image.

#### Ticket Link
[Github Issue 20471](https://github.com/mattermost/mattermost-server/issues/20471)
[MM-44954](https://mattermost.atlassian.net/browse/MM-44954)

#### Release Note

```release-note
Users with default profile pictures will now regenerate a new picture when their username is changed
```

[MM-44954]: https://mattermost.atlassian.net/browse/MM-44954?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by thinkGeist** [comment]
As per the note above this PR has been re-created and as such there are some discussion points on the original PR. I'm going to try and capture the current state/questions in this comment (which is copied from there): 

The main point of discussion on this PR was around the use of negative timestamps to indicate that the set profile picture is in fact a "default" profile picture. I took this approach because it allows the image to automatically update in clients similar to a real profile picture change but to still keep a distinction between a default profile image and a "set" profile image.

One of the suggestions (from @isacikgoz and others) was to use positive timestamps (similar to a real profile picture), but I wanted to raise the fact that this introduces some poor UX around allowing the user to "reset" their profile picture because the positive timestamp causes us to believe that the newly set image can be reset to default now.

For example - if updating a default profile image sets the timestamp to a positive value we get the following behavior as the system "thinks" the profile picture can be reset:

https://user-images.githubusercontent.com/9254315/225048706-fbf91299-d188-443d-a9e5-4343b83d65e3.mp4

In essence I'm just looking for a path forward on this fix, the way I see it there's a couple (an of course feel free to add more that I may be missing):
1. Use negative timestamps to allow us to discern default images from set ones
2. Add a "flag" to the DB somewhere alongside the profile/image to indicate that the profile image is "default"
3. Use positive timestamps and accept the UX issues introduced

**Comment by harshilsharma63** [comment]
@thinkGeist Just to confirm, this change will not update the existing mismatch-ing user's profile picture, right? It fixes the pic when the user next updates their username, right?

**Comment by harshilsharma63** [comment]
@hmhealey since you raised some minor concerns about negative timestamp, what's your final recommendation about it? I think its a nice and clever way and if all tests pass, including E2E, we should be good to go. The negative timestamp will immediately be updated to a positive one anyways as soon as the new profile pic is generated.

Thoughts?

**Review by harshilsharma63 — APPROVED** [review]

**Comment by thinkGeist** [comment]
> @thinkGeist Just to confirm, this change will not update the existing mismatch-ing user's profile picture, right? It fixes the pic when the user next updates their username, right?

@harshilsharma63 Correct this change only impacts future username changes and would not retroactively fix default pictures which are already out of sync.

**Comment by Willyfrog** [comment]
It was @isacikgoz who raised the issue in the other PR but I'm not completely clear on what would be the problem of using a negative timestamp to store the info in the DB. If the problem is transmitting the info "encrypted" it could just be stored that way in the DB and *translated* when adding the info to the struct in go (something like db_timestamp = -X -> {timestamp: x, default: true}, db_timestamp = Y -> {timestamp: Y, default: false}).

I don't think the option of having the X (for dismissing the custom profile pic) available when it is in fact the default profile would be acceptable to UX if there is an easy way to prevent it. So it is either the extra field in the DB or the negative as a way to prevent having an extra field there.

**Comment by Willyfrog** [comment]
After discussing a bit more with @hmhealey we think that it is better to add an extra field in the DB for storing if the picture is the default one or not. If the user's table is too bloated or if we expect any problems from adding a new extra field to the DB, the we can go with the negative timestamp solution. @isacikgoz would that work for server platform?

@thinkGeist would that help for moving forward this PR?

**Comment by thinkGeist** [comment]
@Willyfrog This definitely gives me a path forward, thanks for having that conversation. Mind you it might take me a bit to turn this solution around since I'm unfamiliar with how the server manages DB migrations and the like currently. I will dig into it and report back here once I have an updated solution

**Comment by hmhealey** [comment]
Thanks and sorry for the confusion. I'm not particularly familiar with the current database migration code, but I can definitely try to help point you in the right direction

**Comment by isacikgoz** [comment]
> After discussing a bit more with @hmhealey we think that it is better to add an extra field in the DB for storing if the picture is the default one or not. If the user's table is too bloated or if we expect any problems from adding a new extra field to the DB, the we can go with the negative timestamp solution. @isacikgoz would that work for server platform?

Adding extra column for Users table is not a big issue but I'd try to avoid doing so. Especially for this edge case. Let's "encrypt" this into timestamp if you think this can't be done anyother way.

**Comment by hmhealey** [comment]
Yeah, I think that's our best bet without modifying the User table or making some bigger changes

**Review by hmhealey — APPROVED** [review]

File: server/channels/store/storetest/user_store.go [hunk 1]
```diff
@@ -5909,8 +5910,8 @@ func testUserStoreResetLastPictureUpdate(t *testing.T, ss store.Store) {
 	user, err := ss.User().Get(context.Background(), u1.Id)
 	require.NoError(t, err)
 
-	assert.NotZero(t, user.LastPictureUpdate)
-	assert.NotZero(t, user.UpdateAt)
+	assert.GreaterOrEqual(t, user.LastPictureUpdate, startTime)
+	assert.GreaterOrEqual(t, user.LastPictureUpdate, startTime)
```
Comment (kostaspt) [inline server/channels/store/storetest/user_store.go:5914]: nit: duplicate check

**Review by kostaspt — APPROVED** [review]

File: server/channels/store/storetest/user_store.go (hunk 1 above)
Comment (thinkGeist) [reply server/channels/store/storetest/user_store.go]: Cleaned up! Thanks for pointing it out

**Review by furqanmlk — APPROVED** [review]
Working as expected

Before 

image

After

image

**Pull request merged**